File: test_perf.lua

require 'math'
require 'cutorch'
require 'nn'
require 'cunn'
require 'rnnlib'

local tablex  = require 'pl.tablex'
local stringx = require 'pl.stringx'
local tnt     = require 'torchnet'
local optim   = require 'optim'
local data    = require 'data'
local utils   = require 'utils'
local json    = require 'json'
local os      = require 'os'

local word2vec = require 'utils.word2vec'
local cmd      = torch.CmdLine('-', '-')

cmd:option('-nhid',    256,           'nhid')
cmd:option('-batch',   64,            'batch')
cmd:option('-niter',   1000,          'niter')
cmd:option('-cutoff',  '2000,10000',  'cutoff')
cmd:option('-ntoken',  100000,        'ntoken')

local config = cmd:parse(arg)

local cutoff = tablex.map(tonumber, stringx.split(config.cutoff, ','))
table.insert(cutoff, config.ntoken)

local decoder   = nn.AdaptiveSoftMax(config.nhid, cutoff, config.divval)
local criterion = nn.AdaptiveLoss(cutoff)

local input  = torch.FloatTensor(config.batch * config.niter, config.nhid)
local target = torch.LongTensor(config.batch * config.niter)

input:uniform(-0.1, 0.1)
target:random(1, config.ntoken)

decoder:cuda()
criterion:cuda()
input = input:cuda()
target = target:cuda()

idx = 1

timer = torch.Timer()

while idx < target:size()[1] do
    local x = input:narrow(1, idx, config.batch)
    local y = target:narrow(1, idx, config.batch)

    decoder:zeroGradParameters()

    decoder:setTarget(y)
    decoder:forward(x)
    criterion:forward(decoder.output, y)

    criterion:backward(decoder.output, y)
    decoder:backward(x, criterion.gradInput)

    idx = idx + config.batch
end

print(timer:time().real)

File: test_perf.py
import argparse
import torch 
from torch import nn
from _asm import AdaptiveLogSoftmaxWithLoss
from time import perf_counter


def main(config):
    
    decoder = AdaptiveLogSoftmaxWithLoss(in_features=config.nhid, n_classes=config.ntoken, cutoffs=config.cutoff)

    input  = torch.FloatTensor(config.batch * config.niter, config.nhid).uniform_(-0.1, 0.1)
    target = torch.randint(low=0, high=config.ntoken, size=(config.batch * config.niter,), dtype=torch.long)

    decoder = decoder.to('cuda')
    input   = input.to('cuda')
    target  = target.to('cuda')

    idx = 0
    start = perf_counter()

    while idx < target.size(0):

        x = input.narrow(0, idx, config.batch)
        y = target.narrow(0, idx, config.batch)

        decoder.zero_grad()

        loss = decoder(x, y).loss
        loss.backward()

        idx = idx + config.batch
    
    stop = perf_counter()

    print(stop - start)

if __name__ == '__main__':
    
    parser = argparse.ArgumentParser()
    parser.add_argument('-nhid',    default=256, type=int)
    parser.add_argument('-batch',   default=64, type=int)
    parser.add_argument('-niter',   default=1000, type=int)
    parser.add_argument('-cutoff',  default='2000,10000')
    parser.add_argument('-ntoken',  default=100000, type=int)

    args = parser.parse_args()
    args.cutoff = [int(item) for item in args.cutoff.split(',')]

    main(args)
